contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    Ok(Response::new()
+        .add_attribute("action", "burn")
+        .add_submessage(msg))
+}
+
+fn execute_mint(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+    if token_ids.len() != token_uris.len() {
+        return Err(ContractError::ImbalancedTokenInfo {});
+    }
+    let receiver = deps.api.addr_validate(&receiver)?;
+    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
+
+    let mint_messages = token_ids
+        .into_iter()
+        // Can zip without worrying about dropping data as we assert
+        // that lengths are the same above.
+        .zip(token_uris.into_iter())
+        .map(|(token_id, token_uri)| -> StdResult<WasmMsg> {
+            let msg = cw721_base::msg::ExecuteMsg::Mint(cw721_base::MintMsg::<Empty> {
+                token_id,
+                token_uri: Some(token_uri),
+                owner: receiver.to_string(),
+                extension: Empty::default(),
+            });
+            Ok(WasmMsg::Execute {
+                contract_addr: cw721_addr.to_string(),
+                msg: to_binary(&msg)?,
+                funds: vec![],
+            })
+        })
+        .collect::<StdResult<Vec<_>>>()?;
+
+    Ok(Response::default()
+        .add_attribute("method", "execute_mint")
+        .add_messages(mint_messages))
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_do_instantiate_and_mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    class_uri: Option<String>,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Optionally, instantiate a new cw721 contract if one does not
+    // yet exist.
+    let submessages = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        let expected = CLASS_ID_TO_CLASS_URI.load(deps.storage, class_id.clone())?;
+        if expected != class_uri {
+            // classUri information for a classID shouldn't change
+            // across sends. TODO: is this the case? what are we
+            // suposed to do if not..?
+            return Err(ContractError::ClassUriClash {
+                class_id,
+                expected,
+                actual: class_uri,
+            });
+        }
+        vec![]
+    } else {
+        // Store mapping from classID to classUri. cw721 does not do
+        // this, so we need to do it to stop the infomation from
+        // getting lost.
+        CLASS_ID_TO_CLASS_URI.save(deps.storage, class_id.clone(), &class_uri)?;
+
+        let message = SubMsg::<Empty>::reply_on_success(
```

Comment:
Can we add place keeper link for the diagram that will eventually explain this madness?

Reply:
Started an issue for this: https://github.com/public-awesome/ics721/issues/24

Feel free to add to it as we need more diagram ideas.

Reply:
@JakeHartnell thanks much 🙏